ppci/lang/c/semantics.py

```diff
@@ -320,17 +320,21 @@ def check_redeclaration_storage_class(self, sym, declaration):
         old_storage_class = sym.declaration.storage_class
         new_storage_class = declaration.storage_class
         # None == automatic storage class.
-        invalid_combos = [(None, "static"), ("extern", "static")]
+        # changes of linkage between internal and external are illegal
+        invalid_combos = [(None, "static"),
+                          ("extern", "static"),
+                          ("static", None)]
         combo = (old_storage_class, new_storage_class)
         if combo in invalid_combos:
             message = "Invalid redefine of storage class. Was {}, but now {}".format(
                 old_storage_class, new_storage_class
             )
             self.invalid_redeclaration(sym, declaration, message)
 
-        if not declaration.storage_class:
-            if sym.declaration.storage_class:
-                declaration.storage_class = sym.declaration.storage_class
+        # if new storage-class is "extern", keep the old storage-class
+        # otherwise use the new one
+        if new_storage_class == "extern":
```

Comment:
Rather than changing the storage class, it is probably better to change the position in the Symbol.declarations list of this declaration.